src/mesh.cpp

```diff
@@ -102,6 +154,73 @@ StructuredMesh::bin_label(int bin) const {
   }
 }
 
+//==============================================================================
+// Unstructured Mesh implementation
+//==============================================================================
+
+UnstructuredMesh::UnstructuredMesh(pugi::xml_node node) : Mesh(node) {
+  n_dimension_ = 3;
+
+  // check the mesh type
+  if (check_for_node(node, "type")) {
+    auto temp = get_node_value(node, "type", true, true);
+    if (temp != "unstructured") {
+      fatal_error(fmt::format("Invalid mesh type: {}", temp));
+    }
+  }
+
+  // get the filename of the unstructured mesh to load
+  if (check_for_node(node, "filename")) {
+    filename_ = get_node_value(node, "filename");
+    if (!file_exists(filename_)) {
+      fatal_error("Mesh file '" + filename_ + "' does not exist!");
+    }
+  }
+  else {
+    fatal_error(fmt::format("No filename supplied for unstructured mesh with ID: {}", id_));
+  }
+
+  // check if mesh tally data should be written with
+  // statepoint files
+  if (check_for_node(node, "output")) {
+    output_ = get_node_value_bool(node, "output");
+  }
+
+}
+
+void
+UnstructuredMesh::surface_bins_crossed(const Particle& p,
+                                       std::vector<int>& bins) const {
+  throw std::runtime_error{"Unstructured mesh surface tallies are not implemented."};
+}
+
+std::string
+UnstructuredMesh::bin_label(int bin) const {
+  return fmt::format("Mesh Index ({})", bin);
+};
+
+void
+UnstructuredMesh::to_hdf5(hid_t group) const
+{
+    hid_t mesh_group = create_group(group, fmt::format("mesh {}", id_));
+
+    write_dataset(mesh_group, "type", "unstructured");
+    write_dataset(mesh_group, "filename", filename_);
+    write_dataset(mesh_group, "library", this->library());
+    // write volume of each tet
+    std::vector<double> tet_vols;
+    xt::xtensor<double, 2> centroids({(size_t)this->n_bins(), 3});
+    for (int i = 0; i < this->n_bins(); i++) {
+      tet_vols.emplace_back(this->volume(i));
```

Comment:
I don't have a question here, just a comment - it would be great if we could leverage this volume calculation in Cardinal so we don't have to explicitly tell OpenMC what the volumes are for normalization!

Reply:
If we're using the unstructured mesh tally, we currently do use the element volumes to normalize the tally values. The volumes in the input file are used for the cell-based tally only -- though I do think we could clean streamline that a bit.

Reply:
gotcha, thanks!